(2) Quinn Introduction Chapter, Data transfer (#922)
quinn-rs/quinn: docs/book/src/quinn/data-transfer.md

# Data Transfer

> In the [previous chapter](set-up-connection.md) we characterized how to set up an [Endpoint][Endpoint]

Reviewer: `characterizing` doesn't really make sense in this sentence.

and then get access to a [NewConnection][NewConnection] instance.
Now we will continue with the subject of sending data over this connection.

## Multiplexing

Multiplexing is the act of combining data from multiple streams into a single stream. 
This can have a significant positive effect on the performance of the application. 

> QUIC it itself does not perform multiplexing instead it gives the programmer full control over this process. 

Reviewer: This sentence doesn't make sense to me. QUIC does multiplex frames for all streams into what is conceptually a single stream of QUIC packets.

Author: I mean, like in HTTP/2 multiplexing is performed for you, in QUIC you're responsible for it.

Reviewer: I don't agree. In the Quinn API, the user is certainly not on the hook to merge together parts of different streams into particular packets.

Author: That's right. Maybe I need to rephrase it to the user being in control over how to allocate streams and spread data over them.

  
## Stream Types

QUIC provides support for both stream and message based communication.
Streams and messages can be initiated both on the client and server.

| Type | Description | Reference |
| :----- | :----- | :----- |
| **Bidirectional Stream** | two way stream communication. | see [open_bi][open_bi] |
| **Unidirectional Stream** | one way stream communication. | see [open_uni][open_uni] |
| **Unreliable Messaging (extension)** | message based unreliable communication. | see [send_datagram][send_datagram] |

## How to Use

New streams can be created with the methods [open_bi][open_bi], [open_uni][open_uni] of type [Connection][Connection].
An instance of this type, together with existing streams, can be found in the [connection][connection] field of [NewConnection].

> *Iterate over various streams*

Reviewer: Here and below, I'm not sure `iterate` makes sense to me. What we're really doing here is waiting for incoming streams, which is more like "receiving" than "iterating", perhaps? Also the example code can be misleading. Also maybe the example should maybe make clear that you probably shouldn't use code like this because it would never get to receive bidirectional streams or datagrams, since it'd be waiting for unidirectional streams forever (until the peer closes its connection).

Author: Good one. I mainly added this example to show a very minimal usage of receiving from different streams. I will add a note for that. Also receiving would make more sense yea.

```rust
async fn iterate_streams(mut connection: NewConnection) -> anyhow::Result<()> {
    // Iterate unidirectional streams with only the receiving side.
    while let Some(Ok(recv)) = connection.uni_streams.next().await { }
    // Iterate bidirectional streams with both sent and receiving side.
    while let Some(Ok((sent, recv))) = connection.bi_streams.next().await { }
    // Iterate arrived datagrams.
    while let Some(Ok(bytes)) = connection.datagrams.next().await { }

    Ok(())
}
```

> *Open various kinds streams*

Reviewer: ```suggestion
*Open various kinds of streams*
```

    let mut send = connection.
        open_uni()
        .await?;

    // Open bidirectional stream.
    let (send, recv) = connection.
        open_bi()
        .await?;

    Ok(())
}
```

## Bidirectional Streams

With bidirectional streams data can be sent in both directions, for example, from the connection initiator to the peer and the other way around.
 
*open bidirectional stream*

```rust
async fn open_bidirectional_stream(connection: Connection) -> anyhow::Result<()> {
    let (mut send, recv) = connection.
        open_bi()
        .await?;

    send.write_all(b"test").await?;
    send.finish().await?;
    
    let received = recv.read_to_end(10).await?;

    Ok(())
}
```

*iterate bidirectional stream(s)*

```rust
async fn receive_bidirectional_stream(mut connection: NewConnection) -> anyhow::Result<()> {
    while let Some(Ok((sent, recv))) = connection.bi_streams.next().await {
        // Because it is a bidirectional stream, we can both sent and recieve.
        println!("request: {:?}", recv.read_to_end(50).await?);

        send.write_all(b"response").await?;
        send.finish().await?;
    }

    Ok(())
}
```

## Unidirectional Streams 

> With unidirectional streams, you can carry data only in one direction, for example, from the initiator of the stream to its peer.

Reviewer: ```suggestion
With unidirectional streams, you can carry data only in one direction: from the initiator of the stream to its peer.
```

    
*open unidirectional stream*

```rust
async fn open_unidirectional_stream(connection: Connection)-> anyhow::Result<()> {
    let mut send = connection.
        open_uni()
        .await?;

    send.write_all(b"test").await.unwrap();
    send.finish().await?;

    Ok(())
}
```

*iterating unidirectional stream(s)*

```rust
async fn receive_unidirectional_stream(mut connection: NewConnection) -> anyhow::Result<()> {
    while let Some(Ok(recv)) = connection.uni_streams.next().await {
        // Because it is a unidirectional stream, we can only receive not sent back.
        println!("{:?}", recv.read_to_end(50).await?);
    }

    Ok(())
}
```

## Unreliable Messaging

> With unreliable messaging you can transfer data unreliable UDP.

Reviewer: This isn't really a sentence. Also maybe clarify why unreliability is desirable for some kinds of data, and why datagrams can be a better solution rather than having lots of short unidirectional streams.

Author: What is the benefit for many unidirectional streams vs unreliable messaging?

Reviewer: Many unidirectional streams get you reliability without ordering (so no head-of-line blocking).

Reviewer: Since this is still open, I'm waiting for you to follow up, but perhaps you are waiting for my review? Just realized that it's the end of January and we haven't completely merged your work yet, sorry about that!

Author: This is an old comment and some things have changed already.  Tho I will introduce 'Many unidirectional streams get you reliability without ordering (so no head-of-line blocking)' into this text.

# Data Transfer

> In the [previous chapter](set-up-connection.md) it is explained how to set up an [Endpoint][Endpoint]
> and then get access to a [NewConnection][NewConnection] instance.

Reviewer: ```suggestion
The [previous chapter](set-up-connection.md) explained how to set up an [Endpoint][Endpoint]
and then get access to a [NewConnection][NewConnection] instance.
```

    while let Some(Ok((sent, recv))) = connection.bi_streams.next().await {
        // Because it is a bidirectional stream, we can both sent and recieve.
        println!("request: {:?}", recv.read_to_end(50).await?);

        send.write_all(b"response").await?;
        send.finish().await?;
    }

    Ok(())
}
```

## Unidirectional Streams 

With unidirectional streams, you can carry data only in one direction: from the initiator of the stream to its peer.
    
*open unidirectional stream*

```rust
async fn open_unidirectional_stream(connection: Connection)-> anyhow::Result<()> {
    let mut send = connection.
        open_uni()
        .await?;

    send.write_all(b"test").await.unwrap();
    send.finish().await?;

    Ok(())
}
```

*iterating incoming unidirectional stream(s)*

```rust
async fn receive_unidirectional_stream(mut connection: NewConnection) -> anyhow::Result<()> {
    while let Some(Ok(recv)) = connection.uni_streams.next().await {
        // Because it is a unidirectional stream, we can only receive not sent back.
        println!("{:?}", recv.read_to_end(50).await?);
    }

    Ok(())
}
```

## Unreliable Messaging

> With unreliable messaging you can transfer data over UDP without reliability. 

Reviewer: ```suggestion
With unreliable messaging you can transfer data without reliability. 
```

> This could be useful if data-arrival isnt essential or when needing of high-throughput matters and reliab

Reviewer: Unfinished sentence?

```suggestion
This could be useful if data arrival isn't essential or when needing of high throughput matters and reliab
```

*send datagram*

```rust
async fn sent_unreliable(connection: Connection)-> anyhow::Result<()> {

>     connection.
>         send_datagram(b"test".into())

Reviewer: ```suggestion
    connection
        .send_datagram(b"test".into())
```

        .await?;

    Ok(())
}
```

*iterating datagram stream(s)*

```rust
async fn receive_datagram(mut connection: NewConnection) -> anyhow::Result<()> {

>     while let Some(Ok(receivedBytes)) = connection.datagrams.next().await {

Reviewer: ```suggestion
    while let Some(Ok(received_bytes)) = connection.datagrams.next().await {
```

# Data Transfer

In the [previous chapter](set-up-connection.md) we characterized how to set up an [Endpoint][Endpoint]
and then get access to a [NewConnection][NewConnection] instance.
Now we will continue with the subject of sending data over this connection.

## Multiplexing

> A QUIC stream can be compared to a TCP stream, eventhough if it is based on UDP. 

Reviewer: ```suggestion
A QUIC stream can be compared to a TCP stream, even though if it is based on UDP. 
```

The big difference is that you are not limited to a single stream. 

> You can open multiple streams between two peers, this is also called multiplexing.

Reviewer: This doesn't seem accurate. "multiplexing" is the act of combining data from multiple streams in a single stream, which isn't really what you're referring to here.

Author: Agree, I will rephrase as you posted.

Stream multiplexing can have a significant positive effect on application performance if the resources allocated to streams are properly prioritized.

> Currently multiplexing it is used in a protocol like HTTP/2, but unlike HTTP multiplexing, QUIC does not automatically multiplex.

Reviewer: This comparison doesn't make much sense to me. HTTP isn't really a stream-oriented protocol, it's a request-response protocol.

Author: I am trying to shed light on the comparison between how multiplexing works in HTTP protocol and in quic. In quic it is not done automatically were as in HTTP it, right?

Author: I am going to rewrite this part a bit differently. I think there is some unnecessary info here.

This is an advantage because the user gets full control over the multiplexing. 

> Another difference compared to a TCP stream is that QUIC is not limited to one kind of stream but can open both bidirectional and unidirectional streams.

Reviewer: I think TCP sockets can close one side of the connection, which also results in a unidirectional stream? So this is not quite accurate. I'd just leave the TCP comparison out here and briefly explain that there are four kinds of streams for each QUIC connection: client/server-initiated bi/uni-directional.

Author: Good one!

 
## Stream Types

Quinn offers three ways to send your data. 
Two stream-based and one message-based.

| Type | Description | Reference |
| :----- | :----- | :----- |
| **Bidirectional Stream** | two way stream communication. | see [open_bi][open_bi] |
| **Unidirectional Stream** | one way stream communication. | see [open_uni][open_uni] |

> | **Unreliable Messaging** | message based unreliable communication. | see [send_datagram][send_datagram] |

Reviewer: Maybe should mention that datagrams are an extension, not yet standard, and not required to be implemented.

## How to Use

You can open a new stream or read from an existing stream.
New streams can be created with the methods [open_bi][open_bi], [open_uni][open_uni] from [Connection][Connection] type.
An instance of this type is found in the [connection][connection] field of [NewConnection][NewConnection]. 
This [NewConnection][NewConnection] stores existing streams as well. Lets look at some examples:

*Iterate over various opened streams*

```rust
async fn iterate_streams(mut connection: NewConnection) -> anyhow::Result<()> {
    // Iterate unidirectional streams with only the receiving side.
    while let Some(Ok(recv)) = connection.uni_streams.next().await { }
    // Iterate bidirectional streams with both sent and receiving side.
    while let Some(Ok((sent, recv))) = connection.bi_streams.next().await { }
    // Iterate arrived datagrams.
    while let Some(Ok(bytes)) = connection.datagrams.next().await { }

    Ok(())

> }

Reviewer: "opened" doesn't seem like the best term here, let's use "incoming" instead to make it clear that these were initiated by the peer?

Author: I will just remove this word and leave it at 'iterate over various streams'. Incoming is may not be correct because it can be initiated from the endpoint itself.

```

*Open different types of streams*

```rust
async fn open_streams(mut connection: Connection) -> anyhow::Result<()> {
    // Open unidirectional stream.
    let mut send = connection.
        open_uni()
        .await?;

    // Open bidirectional stream.
    let (send, recv) = connection.
        open_bi()
        .await?;

    Ok(())
}
```

## Bidirectional Streams

> With bidirectional streams you can carry data in both directions, for example, from the initiator to the peer and in reverse.

Reviewer: ```suggestion
With bidirectional streams data can be sent in both directions, for example, from the connection initiator to the peer and the other way around.
```

    while let Some(Ok((sent, recv))) = connection.bi_streams.next().await {
        // Because it is a bidirectional stream, we can both sent and recieve.
        println!("request: {:?}", recv.read_to_end(50).await?);

        send.write_all(b"response").await?;
        send.finish().await?;
    }

    Ok(())
}
```

## Unidirectional Streams 

With unidirectional streams, you can carry data only in one direction, for example, from the initiator of the stream to its peer.
    
*open unidirectional stream*

```rust
async fn open_unidirectional_stream(connection: Connection)-> anyhow::Result<()> {
    let mut send = connection.
        open_uni()
        .await?;

    send.write_all(b"test").await.unwrap();
    send.finish().await?;

    Ok(())
}
```

*iterating unidirectional stream(s)*

```rust
async fn receive_unidirectional_stream(mut connection: NewConnection) -> anyhow::Result<()> {
    while let Some(Ok(recv)) = connection.uni_streams.next().await {
        // Because it is a unidirectional stream, we can only receive not sent back.
        println!("{:?}", recv.read_to_end(50).await?);
    }

    Ok(())
}
```

## Unreliable Messaging

> With unreliable messaging you can transfer data unreliable over bare UDP.

Reviewer: This isn't really "over bare UDP"; datagram frames are embedded in QUIC packets, which are bundled up in UDP datagrams.

Author: Good point. I will mention that instead.
